iothub/service/src/ConnectionLeaseTimeoutHandler.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// HttpClient delegating handler for periodically revoking the connection lease after a specified amount of time. This allows cached HttpClients to periodically do DNS lookups
+    /// which keeps the client communicating with the correct service endpoint in failover scenarios.
+    /// </summary>
+    internal class ConnectionLeaseTimeoutHandler : DelegatingHandler
+    {
+        private readonly Stopwatch _stopwatch;
+        private readonly TimeSpan _leaseTimeout;
+
+        internal ConnectionLeaseTimeoutHandler(TimeSpan leaseTimeout, HttpClientHandler innerHandler) : base()
+        {
+            _stopwatch = new Stopwatch();
+            _stopwatch.Start();
+            _leaseTimeout = leaseTimeout;
+            InnerHandler = innerHandler;
+        }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // By default, singleton/static HttpClient instances will never do a DNS lookup after the first request, which leads to some
+            // bugs such as https://github.com/Azure/azure-iot-sdk-csharp/issues/1865
+
+            // This handler is designed to force the connection to close periodically in order to ensure that httpclients are eventually made aware
+            // of any DNS changes from an IoT Hub failing over.
+            TimeSpan timeSinceConnectionClose = _stopwatch.Elapsed;
+            if (timeSinceConnectionClose >= _leaseTimeout)
+            {
+                request.Headers.ConnectionClose = true;
+                _stopwatch.Restart();
+            }
+
+            return base.SendAsync(request, cancellationToken);
```

Comment:
does `SendAsync()` include implicit open or could we get exceptions saying that the transport layer isn't ready yet?

Reply:
I think you've answered it here: https://github.com/Azure/azure-iot-sdk-csharp/pull/1869#discussion_r605881735, so it is implicit open.